test/Java9andUp/src/org/openj9/test/java9AttachAPI/TargetManager.java

```diff
@@ -0,0 +1,339 @@
+		this.targetId = targetId;
+		this.proc = launchTarget(cmdName, targetId, null, null, appArgs);
+	}
+
+	public TargetManager(String cmdName, String targetId,
+			List<String> vmArgs, List<String> appArgs) {
+		this.targetId = targetId;
+		this.proc = launchTarget(cmdName, targetId, null, vmArgs, appArgs);
+	}
+
+	public TargetManager(String cmdName, String targetId, String displayName,
+			List<String> vmArgs, List<String> appArgs) {
+		this.targetId = targetId;
+		this.proc = launchTarget(cmdName, targetId, displayName, vmArgs,
+				appArgs);
+	}
+
+	int waitFor() throws InterruptedException {
+		return proc.waitFor();
+	}
+
+	/**
+	 * launch a TargetVM process
+	 * 
+	 * @return launched process
+	 */
+	private Process launchTarget(String cmdName, String tgtId,
+			String myDisplayName, List<String> vmArgs,
+			List<String> appArgs) {
+		List<String> argBuffer = new ArrayList<String>();
+		String[] args = {};
+		this.displayName = myDisplayName;
+		Runtime me = Runtime.getRuntime();
+		char fs = File.separatorChar;
+		String javaExec = System.getProperty("java.home") + fs + "bin" + fs
+				+ "java";
+		logger.debug("javaExec="+javaExec);
+		String sideCar = System.getProperty("java.sidecar");
+		String myClasspath = System.getProperty("java.class.path");
+		argBuffer.add(javaExec);
+		argBuffer.add("-Dcom.ibm.tools.attach.enable=yes");
+		if (doLogging) {
+			argBuffer.add("-Dcom.ibm.tools.attach.logging=yes");
+		}
+		if (null != tgtId) {
+			argBuffer.add("-Dcom.ibm.tools.attach.id=" + tgtId);
+		}
+		if (null != myDisplayName) {
+			argBuffer.add("-Dcom.ibm.tools.attach.displayName=" + myDisplayName);
+		}
+		argBuffer.add("-classpath");
+		argBuffer.add(myClasspath);
+		if ((null != sideCar) && (sideCar.length() > 0)) {
+			String sidecarArgs[] = sideCar.split(" +");
+			for (String s : sidecarArgs) {
+				argBuffer.add(s);
+			}
+		}
+		if (null != vmArgs) {
+			argBuffer.addAll(vmArgs);
+		}
+		argBuffer.add(cmdName);
+		if (null != appArgs) {
+			argBuffer.addAll(appArgs);
+		}
+		Process target = null;
+		args = new String[argBuffer.size()];
+		argBuffer.toArray(args);
+		if (verbose) {
+			System.out.print("\n");
+			for (int i = 0; i < args.length; ++i) {
+				System.out.print(args[i] + " ");
+			}
+			System.out.print("\n");
+		}
+		try {
+			target = me.exec(args);
+		} catch (IOException e) {
+			e.printStackTrace();
+			return null;
+		}
+		this.targetIn = target.getOutputStream();
+		InputStream targetOut = target.getInputStream();
+		InputStream targetErr = target.getErrorStream();
+		targetInWriter = new BufferedWriter(new OutputStreamWriter(targetIn));
+		targetOutReader = new BufferedReader(new InputStreamReader(targetOut));
+		targetErrReader = new BufferedReader(new InputStreamReader(targetErr));
+		try {
+			targetInWriter.write(TargetManager.TARGETVM_START + '\n');
+			targetInWriter.flush();
+		} catch (IOException e) {
+			e.printStackTrace();
+			return null;
+		}
+		return target;
+	}
+
+	static void listIpcDir() {
```

Comment:
Can you add javadoc for each of these methods?  It looks like this method tries to recursively list all the files and directories under the java.io.tmpdir to the logger.debug() output.  Should it only be called if the logger.debug() level is enabled?

Reply:
> Can you add javadoc for each of these methods? 
Done.  Let me know if you want more explanation.

>It looks like this method tries to recursively list all the files and directories under the java.io.tmpdir to the logger.debug() output. Should it only be called if the logger.debug() level is enabled?
Actually, it lists the contents of /tmp/.com_ibm_tools_attach, which will be a few files for each Java process.  This method is mostly for post-failure analysis.